### ISSUE TITLE
Improved plugin/extension writing guide

### ISSUE THREAD

**Issue by matthewp** [stealjs/stealjs#83]
__tldr;__ update the plugin guide to focus on *how* to write plugins more. Give visualizations where possible to explain the lifecycle hooks.

# The Problem

Many people ask for plugins for various formats that steal doesn't support out of the box. Few people know how to write these plugins. We have a plugin guide which shows an example of creating a plugin, but it doesn't give 

# The Solution

Revamp the plugin guide to give as much detail as possible to make people confident about writing plugins. Steal ideas from [webpack docs](https://webpack.js.org/contribute/writing-a-plugin/) where they are good.

In essence we need to get across:

* The lifecycle hooks and where they fit.
* Focus on __translate__ as that is the most commonly used hook.
* The `!` syntax and where it fits in.
  * The `ext` config and what that is for.
  * The idea of `plugins` configuration, and documenting to your users that they need it.
* That writing plugins is easy and fun.

Additionally we should explore if there are any test tooling that would help. Internally steal uses a project called [steal-test-helpers](https://github.com/stealjs/steal-test-helpers) that makes it easier to write tests for plugins (used heavily by the npm plugin), maybe we should also document and promote this module.

Here's an example test that uses *steal-test-helpers*:

```js
QUnit.test("process.cwd()", function(assert){
	var done = assert.async();

	var appModule = "module.exports = process.cwd();";

	var loader = helpers.clone()
		.rootPackage({
			name: "app",
			main: "main.js",
			version: "1.0.0"
		})
		.withModule("app@1.0.0#main", appModule)
		.loader;

	loader["import"]("app")
	.then(function(app){
		assert.equal(typeof app, "string", "process.cwd is a string");
		assert.equal(app[app.length - 1], "/", "ends in a slash");
	})
	.then(done, done);
});
```

# Tasks

* [ ] - Outline what the plugin guide should be and pass it around to a few people like @pYr0x and @m-mujica for feedback.
* [ ] - See about exporting steal-test-helpers as a package that the community can use for writing plugin tests.
